### PR TITLE
rpc/ethapi: RPCHeader/Block/Withdrawal add `MarshalFastJSONTo`

### PR DESCRIPTION
**Pull request by AskAlexSharov**
`eth_getBlockByNumber(false)` 47.7k -> 99.0k rps (2.08x), p99 2.37 -> 0.25ms. `eth_getHeaderByNumber` 160.0k -> 174.4k (+9%). Blocks and headers byte-identical. 5 ABBA rounds, rounds within 0.7%. Needs #24127 (concrete slice types) for the hash fast path, and #24105 for the JSONWriter verbs.

Out of scope: 
- txs, uncles, etc... 
- changing RPCHeader classes

### CODE REVIEW

File: rpc/ethapi/header_fastjson.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+		switch txs := b.Transactions.(type) {
+		case []common.Hash:
+			f.field("transactions").WriteArrayStart()
+			for i := range txs {
+				if i > 0 {
+					w.WriteMore()
+				}
+				w.WriteHex(txs[i][:])
+			}
+			w.WriteArrayEnd()
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply rpc/ethapi/header_fastjson.go]: Fixed: writeHashArray writes null for a nil slice and [] for an empty one, for both transactions and uncles. Pinned by the typed-nil and nil-uncles cases in TestRPCBlockMarshalFastJSONTo, which fail without the guard.

File: rpc/ethapi/block_fastjson.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+		for i := range hashes {
+			if i > 0 {
+				w.WriteMore()
+			}
+			w.WriteHex(hashes[i][:])
+		}
+		w.WriteArrayEnd()
+	case fullTxs != nil:
+		field(w, "transactions").WriteRawBytes(fullTxs)
+	}
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply rpc/ethapi/block_fastjson.go:167]: Fixed: same writeHashArray guard covers this branch.

File: rpc/jsonrpc/eth_api.go [hunk 3]
```diff
@@ -180,6 +181,10 @@ type BaseAPI struct {
 	witnessCache *witnessResultCache
 }
 
+// BlockCacheBytes bounds the decoded blocks the RPC layer keeps. A mainnet block costs ~331KB of
+// heap, so this holds ~1600 of them, about 5 hours of chain.
+var BlockCacheBytes = dbg.EnvDataSize("RPC_BLOCK_CACHE", 512*datasize.MB)
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply rpc/jsonrpc/eth_api.go:185]: Not this PR: `rpc/jsonrpc/eth_api.go` is not in the diff. The 512MB block cache came from #24120, already on main.

File: rpc/jsonrpc/eth_api.go [hunk 4]
```diff
@@ -21,6 +21,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/erigontech/erigon/common/dbg"
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply rpc/jsonrpc/eth_api.go:24]: Same as above: `rpc/jsonrpc/eth_api.go` is not in this diff, that import group is #24120 on main.

File: rpc/ethapi/block_fastjson.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+		w.WriteHex(hashes[i][:])
+	}
+	w.WriteArrayEnd()
+}
+
+// writeWithdrawals writes the withdrawal list. Four fixed-width fields each, so it needs
+// no reflection; encoding it here keeps execution/types free of a JSON dependency.
+func writeWithdrawals(w jsonw.JSONWriter, name string, ws *types.Withdrawals) {
+	field(w, name).WriteArrayStart()
+	for i, wd := range *ws {
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply rpc/ethapi/block_fastjson.go]: Fixed. `jsonw.Array` takes a pointer to the slice: nil pointer omits the field, nil slice writes `null`, empty slice writes `[]`. Covered by the "withdrawals pointer to nil slice" case.

File: rpc/ethapi/api_test.go [hunk 6]
```diff
@@ -3,6 +3,9 @@ package ethapi
 import (
 	"bytes"
 	"encoding/json"
+	"github.com/erigontech/erigon/common/hexutil"
+	"github.com/erigontech/erigon/rpc/jsonstream"
+	"github.com/erigontech/erigon/rpc/jsonstream/jsonw"
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply rpc/ethapi/api_test.go]: Fixed, the project imports are their own group now.

**Review by taratorio — APPROVED** [review]

File: rpc/jsonstream/stack_stream.go [hunk 7]
```diff
@@ -104,6 +107,70 @@ func (s *StackStream) WriteHex(b []byte) {
...
+
+// HexesField writes a hash array as one field: one buffer growth for the whole array,
+// where a value write per element grows once per hash.
+func HexesField(w jsonw.JSONWriter, name string, hashes []common.Hash) {
+	jsonw.Field(w, name)
+	if hashes == nil {
+		w.WriteNil()
+		return
+	}
+	WriteHexes(Concrete(w), hashes)
```
Comment (awskii) [inline rpc/jsonstream/stack_stream.go:134]: `Concrete` falls through to `return nil` for anything that is not `*StackStream` or `*LazyFieldStream` (:122), and `WriteHexes` dereferences it immediately at :142 (`s.stream.Buffer()`).

Unreachable today — those are the only two `jsonw.JSONWriter` implementations in the tree. But `Stream` is an interface built to be wrapped, and `LazyFieldStream` is already one such wrapper; a second one (tee, limiter, test double) makes `eth_getBlockByNumber` panic on every block with transactions or uncles, since `RPCBlock.MarshalFastJSONTo` routes both fields through `HexesField`.

Cheapest guard is a panic in `Concrete` naming the type, rather than an untested fallback path.

File: rpc/jsonstream/stack_stream.go [hunk 8]
```diff
@@ -104,6 +107,70 @@ func (s *StackStream) WriteHex(b []byte) {
...
+		}
+		buf = hexutil.AppendQuoted(buf, items[i][:])
+	}
+	s.stream.SetBuffer(append(buf, ']'))
+	s.popCommaOrField()
+}
+
+// WriteHexBytes is WriteHexes for elements that are already byte slices, whose lengths vary
+// and so are summed before the single growth.
+func WriteHexBytes[S ~[]E, E ~[]byte](s *StackStream, items S) {
```
Comment (awskii) [inline rpc/jsonstream/stack_stream.go:157]: `WriteHexBytes` is new and exported, and nothing calls it — `grep -rn WriteHexBytes` matches only this definition. No test either, and being exported keeps it out of unused-symbol linting.

That leaves the hand-rolled sizing at :158-161 (`size := 2 + len(items)`, then `hexutil.QuotedLen` per element) with no coverage — the part that differs from `WriteHexes`, which is reached through `HexesField` and is covered by the new byte-equality tests.

Drop it until a caller needs it, or land it with that caller and a test over nil / empty slice / empty element / multi-element.

File: rpc/ethapi/block_fastjson.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+import (
+	"encoding/json"
+
+	"github.com/erigontech/erigon/common"
+	"github.com/erigontech/erigon/execution/types"
+	"github.com/erigontech/erigon/rpc/jsonstream"
+	"github.com/erigontech/erigon/rpc/jsonstream/jsonw"
+)
+
+// MarshalFastJSONTo writes the header as its own object, for eth_getHeaderByNumber and
```
Comment (awskii) [inline rpc/ethapi/block_fastjson.go]: The comment names its two callers ("for eth_getHeaderByNumber and eth_getHeaderByHash"). A third method returning `*RPCHeader`, or a rename of either, leaves it quietly wrong, and `grep RPCHeader` already answers the question — call-site lists belong in the PR description.

b47bdd052ba in this same PR took a consumer name out of the neighbouring comment in api.go, so this is the one spot the decision didn't reach.

The second sentence is the part a reader can't get from the signature; worth keeping.

**Review by awskii — APPROVED** [review]
Three minors, none blocking.

File: rpc/jsonstream/stack_stream.go (hunk 7 above)
Comment (AskAlexSharov) [reply rpc/jsonstream/stack_stream.go:134]: Done: `Concrete` panics with the writer's type instead of returning nil.

File: rpc/jsonstream/stack_stream.go (hunk 8 above)
Comment (AskAlexSharov) [reply rpc/jsonstream/stack_stream.go:157]: Its caller landed on this branch with #24141 (`account_proof.go`, getProof). Added `TestWriteHexBytes` over nil, empty, an empty element and several elements.

File: rpc/ethapi/block_fastjson.go (hunk 9 above)
Comment (AskAlexSharov) [reply rpc/ethapi/block_fastjson.go]: Done: dropped the caller names, kept the second sentence.
